### Ejercicio 1.
Extraer los siguentes estadísiticos del conjunto de datos para varias de las varibles disponibles:
* Tabla de contingencia
* Coeficiente de correlación de las variables numéricas del dataset.
* Sesgos de alguna de las variables numéricas del dataset.
* Intervalo de confianza de algunas de las variables numéricas del dataset

In [1]:
import pandas as pd
import numpy as np
from scipy.stats import skew
import scipy.stats as st
import re

In [2]:
df_influencers = pd.read_csv('../Estadistica/instagram_global_top_1000.csv')
df_influencers.head(3)

Country  Rank      Account              Title  \
0     All     1    cristiano  Cristiano Ronaldo   
1     All     2  kyliejenner            Kylie 🤍   
2     All     3     leomessi          Leo Messi   

                                     Link                   Category  \
0    https://www.instagram.com/cristiano/         Sports with a ball   
1  https://www.instagram.com/kyliejenner/    Fashion|Modeling|Beauty   
2     https://www.instagram.com/leomessi/  Sports with a ball|Family   

     Followers Audience Country  Authentic engagement  Engagement avg  \
0  400100000.0            India             7800000.0       9500000.0   
1  308800000.0    United States             6200000.0      10100000.0   
2  306300000.0        Argentina             4800000.0       6500000.0   

                      Scraped  
0  2022-02-07 16:50:24.798803  
1  2022-02-07 16:50:24.798803  
2  2022-02-07 16:50:24.798803

In [3]:
#exploramos las columnas y el tipo de datos que contienen
df_influencers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country               1000 non-null   object 
 1   Rank                  1000 non-null   int64  
 2   Account               1000 non-null   object 
 3   Title                 983 non-null    object 
 4   Link                  1000 non-null   object 
 5   Category              909 non-null    object 
 6   Followers             1000 non-null   float64
 7   Audience Country      993 non-null    object 
 8   Authentic engagement  1000 non-null   float64
 9   Engagement avg        1000 non-null   float64
 10  Scraped               1000 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 86.1+ KB


In [3]:
df_influencers['Category'].isnull().sum()

91

In [86]:
df_influencers['Category'].dropna()

0                       Sports with a ball
1                  Fashion|Modeling|Beauty
2                Sports with a ball|Family
3                         Modeling|Fashion
4                          Music|Lifestyle
                      ...                 
995               Lifestyle|Music|Modeling
996                     Sports with a ball
997                            Fitness|Gym
998    Art|Artists|Cinema|Actors/actresses
999                              Lifestyle
Name: Category, Length: 909, dtype: object

In [30]:
df_influencers.describe()

Rank     Followers  Authentic engagement  Engagement avg
count  1000.000000  1.000000e+03          1.000000e+03    1.000000e+03
mean    500.500000  2.483610e+07          6.491228e+05    8.653886e+05
std     288.819436  3.872143e+07          1.098804e+06    1.267515e+06
min       1.000000  2.800000e+06          0.000000e+00    5.200000e+04
25%     250.750000  9.000000e+06          2.091000e+05    3.086000e+05
50%     500.500000  1.440000e+07          3.637000e+05    5.239000e+05
75%     750.250000  2.565000e+07          6.832500e+05    9.508000e+05
max    1000.000000  4.696000e+08          1.340000e+07    1.340000e+07

In [32]:
df_influencers.isnull().sum()

Country                  0
Rank                     0
Account                  0
Title                   17
Link                     0
Category                91
Followers                0
Audience Country         7
Authentic engagement     0
Engagement avg           0
Scraped                  0
dtype: int64

     Vamos a crear una nueva columna llamada 'Continentes' para poder facilitar la legibilidad de la tabla de contingencias entre Países de audiencia x Tipo de contenido. Para ello:
     * Creamos la función unir_paises
     * La aplicamos a la creación de la nueva columna.

In [33]:
#creación de la función
def unir_paises (col):
    EEUU = ['United States']
    Rusia = ['Russia']
    Latam = ['Argentina', 'Brazil', 'Mexico', 'Colombia', 'Chile']
    UE = ['France', 'Spain', 'Italy', 'Poland', 'Germany']
    Asia_SE = ['Indonesia', 'Japan', 'South Korea', 'Thailand', 'Philippines']
    Africa = ['Nigeria', 'Morocco', 'Algeria']
    Asia_arabe = ['Iran', 'Egypt', 'Turkey', 'Saudi Arabia', 'United Arab Emirates', 'Iraq', 'Syria', 'Kazakhstan']

    if col in Asia_SE:
        return 'Sudeste asiático'
    elif col in UE:
        return 'UE'
    elif col in Latam:
        return 'Latam'
    elif col in Africa:
        return 'Africa'
    elif col in Asia_arabe:
        return 'Países arabes'
    elif col in EEUU:
        return 'EEUU'
    elif col in Rusia:
        return 'Russia'

In [34]:
#aplicación de la función a la nueva columna
df_influencers['Continentes'] = df_influencers['Audience Country'].apply(unir_paises)

In [35]:
#comprobación de la correcta creación de la columna
df_influencers['Continentes'].unique()

array([None, 'EEUU', 'Latam', 'Sudeste asiático', 'Países arabes', 'UE',
       'Russia', 'Africa'], dtype=object)

In [5]:
def clasif_contenido(col):
    Moda = ['Fashion', 'Accessories', 'Beauty', 'Modeling','Clothing','Outfits']
    Est_de_vida = ['Lifestyle']
    Arte = ['Arts', 'Cinema', 'Photography', 'Shows', 'Actors/actresses','Artists']
    Tecnologias = ['Computers','Gadgets','Machinery','Technologies']
    Inversion = ['Finance'', ''Economics','Business','Careers']
    Deportes = '[S|sports]'

    if col in Deportes:
        return 'Deportes'
    elif col in Moda:
        return 'Moda'
    elif col in Est_de_vida:
        return 'Estilo de vida'
    elif col in Arte:
        return 'Arte'
    elif col in Tecnologias:
        return 'Tecnologías'
    elif col in Inversion:
        return 'Inversión'
    else:
        return 'Otros'

In [36]:
df_influencers_crosstab = pd.DataFrame(pd.crosstab(df_influencers['Continentes'], df_influencers['Category']))
df_influencers_crosstab

Category          Accessories|Jewellery|Clothing|Outfits|Fashion  \
Continentes                                                        
Africa                                                         0   
EEUU                                                           1   
Latam                                                          0   
Países arabes                                                  0   
Russia                                                         0   
Sudeste asiático                                               0   
UE                                                             0   

Category          Adult content  Art|Artists  Art|Artists|Beauty  \
Continentes                                                        
Africa                        0            0                   0   
EEUU                          4            1                   1   
Latam                         1            1                   0   
Países arabes                 0            0                   0   
Russia                        0            0                   0   
Sudeste asiático              0            1                   0   
UE                            1            0                   0   

Category          Art|Artists|Cinema|Actors/actresses|Shows  \
Continentes                                                   
Africa                                                    0   
EEUU                                                      1   
Latam                                                     0   
Países arabes                                             0   
Russia                                                    0   
Sudeste asiático                                          0   
UE                                                        0   

Category          Art|Artists|Comics  sketches  Art|Artists|Fashion|Modeling  \
Continentes                                                                    
Africa                                       0                             0   
EEUU                                         1                             1   
Latam                                        0                             1   
Países arabes                                0                             0   
Russia                                       0                             0   
Sudeste asiático                             0                             0   
UE                                           0                             0   

Category          Art|Artists|Fashion|Music  Art|Artists|Shows  Beauty  ...  \
Continentes                                                             ...   
Africa                                    0                  0       0  ...   
EEUU                                      1                  0       4  ...   
Latam                                     0                  0       5  ...   
Países arabes                             0                  0       1  ...   
Russia                                    0                  0       1  ...   
Sudeste asiático                          0                  1       2  ...   
UE                                        0                  0       2  ...   

Category          Shows|Fitness|Gym  Shows|Humor|Fun|Happiness  \
Continentes                                                      
Africa                            0                          0   
EEUU                              1                          4   
Latam                             0                          1   
Países arabes                     0                          0   
Russia                            0                          0   
Sudeste asiático                  0                          1   
UE                                0                          0   

Category          Shows|Lifestyle  Shows|Modeling  Sports with a ball  \
Continentes                                                             
Africa                          

In [6]:
df_influencers['Cat_clasif'] = df_influencers['Category'].apply(clasif_contenido)

TypeError: 'in <string>' requires string as left operand, not float

In [7]:
#matriz de correlación
correlacion_influs = df_influencers.corr()
correlacion_influs

C:\Users\PcCom\AppData\Local\Temp\ipykernel_10204\3838001811.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  correlacion_influs = df_influencers.corr()


Rank  Followers  Authentic engagement  \
Rank                  1.000000  -0.435286             -0.463701   
Followers            -0.435286   1.000000              0.287067   
Authentic engagement -0.463701   0.287067              1.000000   
Engagement avg       -0.506682   0.336169              0.981982   

                      Engagement avg  
Rank                       -0.506682  
Followers                   0.336169  
Authentic engagement        0.981982  
Engagement avg              1.000000

* Intervalo de confianza:

    Vamos a calcular el intervalo de confianza de la variable 'Engagement auténtico'.

In [45]:
datos = df_influencers['Authentic engagement'] 

In [55]:
#Use of keyword argument `alpha` for method `interval` is deprecated. Use first positional argument or keyword argument `confidence` instead.
IC = st.t.interval(confidence=0.99, df=len(df_influencers)-1, loc=np.mean(datos), scale=st.sem(datos))
print(f'El valor que no conocemos de Authentic Engagement está entre {round(IC[0], 2)} e {round(IC[1], 2)} con una confianza del 99%')

El valor que no conocemos de Authentic Engagement está entre 559448.58 e 738797.02 con una confianza del 99%


### Ejercicio 2.

Interpretación de los resultados.

* Ahora interpreta los resultados obtenidos de los diferentes estadísticos estudiados haciendo hincapié en aspectos interesantes de tus datos y explica lo mejor posible la naturaleza de los mismos.

*Matriz de correlación:*

Observamos que existe una correlación positiva muy fuerte (0.98) entre las variables 'Authentic Engagement' y 'Engagement avg', lo cual nos indica que, a mayor interacción con el contenido, mayor cantidad de engagement real.

Existe también una correlación inversa moderada-alta (-0.50) entre la posición en el ranking y la media del engagement, por lo que a mayor posición en el ranking, menor será el engagement promedio.